### PR TITLE
Copter: add "high throttle cancels landing" option

### PR DESCRIPTION
**Pull request by jschall**

### CODE REVIEW

**Comment by magicrub** [comment]
fyi, plane has this via param LAND_ABORT_THR used here:
https://github.com/diydrones/ardupilot/blob/master/ArduPlane/ArduPlane.cpp#L838

**Comment by R-Lefebvre** [comment]
Great feature.

Why not change to Loiter mode instead of Alt_Hold if we have good GPS?

**Comment by OXINARF** [comment]
I think you have seen it wrong Rob, it is changing to Loiter and if it can't then it goes to AltHold

**Comment by R-Lefebvre** [comment]
Ah, yes, thank you I see it now.  Honestly, I find that slightly unclear, setting a mode within an if statement.  I know it's valid, it's just not obvious what's happening there.

File: ArduCopter/control_auto.cpp [hunk 1]
```diff
@@ -324,6 +324,13 @@ void Copter::auto_land_run()
 
     // process pilot's input
     if (!failsafe.radio) {
+        if((g.throttle_behavior & THR_BEHAVE_HIGH_THROTTLE_CANCELS_LAND) != 0 && rc_throttle_control_in_filter.get() > 0.7f*THR_MAX){
+            // exit land if throttle is high
+            if (!set_mode(LOITER)) {
```
Comment (magicrub) [inline ArduCopter/control_auto.cpp]: For clarity, this should read:
bool success = set_mode(LOITER);
if (!success) { set_mode(ALT_HOLD); }

Comment (WickedShell) [inline ArduCopter/control_auto.cpp]: Might be a style thing but I honestly find that less clear.

**Comment by jschall** [comment]
Really? I think we use that convention in other places.

**Comment by R-Lefebvre** [comment]
Which one, the one that magicrub just suggested?

I like what he did there.  It is a little busier, but it at least indicates that something is going on.

if (!do_something()){
   //then
    do_something_else();
}

is not at all clear what is happening.

**Comment by WickedShell** [comment]
I find tom's suggestion less clear personally. It requires me to track down bool's from different lines and makes it easier to accidentally split the logic later. Doing the functional work in the if is clear in that it is a very clear when reading exactly what the dependency is, and makes it harder to separate by accident. The other win is that it removes the implication we make further decisions outside that if block based on the success value. (This may be because I've been doing a lot of functional programming recently). Just my opinion though, I think we do both styles at the moment.

**Comment by billbonney** [comment]
I think what was written by Jon is perfectly readable. The only time I would raise concern with the use of ! is when something is written like

```
if (!do_something() < 0.7f){
   //then
   do_something_else();
}
```
since using parentheses clears up the ambiguity 
```
if (!(do_something() < 0.7f)){
   //then
   do_something_else();
}
```

**Comment by rmackay9** [comment]
merged, thanks!
I added a couple of changes on top including:
1. parameter bitmask to make setting the PILOT_THR_BHV from the GCS easier (if using a GCS that parses the BitMask param value)
2. added logging to dataflash if the pilot cancels the land (this might save some time when doing support and trying to figure out why the mode changed)
3. changed no-gps-land to only switch to AltHold instead of first trying Loiter.  We could change this back so that it tries Loiter if people feel strongly about it.  My concern was just that if a pilot was flying in Loiter with throttle high and an EKF failsafe occurred, the vehicle would first go to non-gps-land but then immediate try to get back into Loiter.  This switch back to Loiter would fail because position_ok would be false but the near circular dependency that could appear through a small future code change worried me slightly.

**Pull request closed without merging**